sstables/sstable_set_impl.hh

```diff
@@ -125,9 +125,16 @@ public:
 // the managed sets cannot be modified through compound_sstable_set, but only jointly read from, so insert() and erase() are disabled.
 class compound_sstable_set : public sstable_set_impl {
     schema_ptr _schema;
-    std::vector<sstable_set> _sets;
+    std::vector<sstable_set> _cloned_sets;
+    std::vector<const sstable_set*> _sets;
+
+    compound_sstable_set(schema_ptr schema, std::vector<sstable_set> cloned_sets);
 public:
-    compound_sstable_set(schema_ptr schema, std::vector<sstable_set> sets);
+    compound_sstable_set(schema_ptr schema, std::vector<const sstable_set*> sets);
+    compound_sstable_set(compound_sstable_set&&) = default;
+    compound_sstable_set(const compound_sstable_set&) = delete;
+    compound_sstable_set& operator=(compound_sstable_set&&) = default;
+    compound_sstable_set& operator=(const compound_sstable_set&) = delete;
 
```

Comment:
this is a very interesting series, but I confess I need to come back tomorrow with a fresher mind. these are big changes. were there any lifetime management issue that worried you at any point in time? how do you guarantee immutability of sstable set in range scan? We cannot clone the set on every read, but rather we prefer to do a copy on write (which frequency is memtable flush). how does incremental_reader_selector keep a immutable version of sstable set without being cloned if table can now update the set in place?

Reply:
> this is a very interesting series, but I confess I need to come back tomorrow with a fresher mind. these are big changes. were there any lifetime management issue that worried you at any point in time? how do you guarantee immutability of sstable set in range scan?

That needs to be fixed in the next revision.

> We cannot clone the set on every read, but rather we prefer to do a copy on write (which frequency is memtable flush). how does incremental_reader_selector keep a immutable version of sstable set without being cloned if table can now update the set in place?

We need to keep a shared pointer of the sstable_set_impl instead of the sstable_set, since updating the sstable_set replaces the shared point it holds, so the previous shared pointer is immutable and holds a snapshot of the sstable_set contents.